plugins/RandomTriggerCandidateMaker.cpp

```diff
@@ -0,0 +1,171 @@
+}
+
+void
+RandomTriggerCandidateMaker::get_info(opmonlib::InfoCollector& /*ci*/, int /*level*/)
+{
+}
+
+void
+RandomTriggerCandidateMaker::do_configure(const nlohmann::json& obj)
+{
+  m_conf = obj.get<randomtriggercandidatemaker::ConfParams>();
+}
+
+void
+RandomTriggerCandidateMaker::do_start(const nlohmann::json& obj)
+{
+  m_run_number = obj.value<dunedaq::dataformats::run_number_t>("run", 0);
+
+  m_running_flag.store(true);
+
+  switch (m_conf.timestamp_method) {
+    case randomtriggercandidatemaker::timestamp_estimation::kTimeSync:
+      TLOG_DEBUG(0) << "Creating TimestampEstimator";
+      m_timestamp_estimator.reset(new TimestampEstimator(m_time_sync_source, m_conf.clock_frequency_hz));
+      break;
+    case randomtriggercandidatemaker::timestamp_estimation::kSystemClock:
+      TLOG_DEBUG(0) << "Creating TimestampEstimatorSystem";
+      m_timestamp_estimator.reset(new TimestampEstimatorSystem(m_conf.clock_frequency_hz));
+      break;
+  }
+
+  m_send_trigger_candidates_thread = std::thread(&RandomTriggerCandidateMaker::send_trigger_candidates, this);
+  pthread_setname_np(m_send_trigger_candidates_thread.native_handle(), "random-tc-maker");
+}
+
+void
+RandomTriggerCandidateMaker::do_stop(const nlohmann::json& /*obj*/)
+{
+  m_running_flag.store(false);
+
+  m_send_trigger_candidates_thread.join();
+
+  m_timestamp_estimator.reset(nullptr); // Calls TimestampEstimator dtor
+}
+
+void
+RandomTriggerCandidateMaker::do_scrap(const nlohmann::json& /*obj*/)
+{
+  m_configured_flag.store(false);
+}
+
+triggeralgs::TriggerCandidate
+RandomTriggerCandidateMaker::create_candidate(dfmessages::timestamp_t timestamp)
+{
+  triggeralgs::TriggerCandidate candidate;
+  candidate.time_start = timestamp;
+  candidate.time_end = timestamp;
+  candidate.time_candidate = timestamp;
+  candidate.detid = { 0 };
+  candidate.type = triggeralgs::TriggerCandidateType::kRandom;
+  candidate.algorithm = 0;
+  candidate.version = 0;
+
+  return candidate;
+}
+
+int
+RandomTriggerCandidateMaker::get_interval(std::mt19937 &gen)
+{
+  switch (m_conf.time_distribution) {
+    case randomtriggercandidatemaker::distribution_type::kUniform:
+      return m_conf.trigger_interval_ticks;
+    case randomtriggercandidatemaker::distribution_type::kPoisson:
+      std::exponential_distribution<double> d(1.0/m_conf.trigger_interval_ticks);
+      return (int)(0.5+d(gen));
+  }
+}
+
+void
+RandomTriggerCandidateMaker::send_trigger_candidates()
+{
+  std::mt19937 gen(m_run_number);
+  // Wait for there to be a valid timestamp estimate before we start
+  if (m_timestamp_estimator->wait_for_valid_timestamp(m_running_flag) == TimestampEstimatorBase::kInterrupted) {
+    return;
+  }
+
+  dfmessages::timestamp_t initial_timestamp = m_timestamp_estimator->get_timestamp_estimate();
+  dfmessages::timestamp_t next_trigger_timestamp = initial_timestamp + get_interval(gen);
```

Comment:
What do you think of rounding the `next_trigger_timestamp` to be divisible by `trigger_interval_ticks` when a uniform interval is used? It's not really a functional change, but it does make the timestamps easier to look at (and to see the interval between maybe?)

Reply:
No strong feelings here. Agreed this is cleaner at a glance for uniform triggers, and adding this shouldn't have a significant impact for other types of triggers.

Reply:
Thanks, that would be a good change